### PR TITLE
Clustering

### PR DESCRIPTION
**Pull request by hntd187**

### CODE REVIEW

File: src/handlers/search.rs [hunk 1]
```diff
@@ -64,36 +83,45 @@ pub mod tests {
...
+            .map(|q| {
+                dbg!(&q);
+                assert_eq!(q.hits, 3);
+            })
+            .map_err(|_| ())
+            .wait()
+            .unwrap();
     }
 
+    // Should I do this? with the tokio::run
```
Comment (mauri870) [inline src/handlers/search.rs]: I don't think that this matters at all in tests, maybe it's better to keep a standard by using tokio::run everywhere.

File: src/cluster/remote_handle.rs [hunk 2]
```diff
@@ -39,25 +59,31 @@ impl IndexHandle for RemoteIndex {
...
+                query: bytes,
             });
+            client
+                .search_index(req)
+                .map(|res| {
+                    info!("RESPONSE = {:?}", res);
+                    res.into_inner()
+                })
+                .map_err(|e| {
+                    info!("ERR = {:?}", e);
```
Comment (LucioFranco) [inline src/cluster/remote_handle.rs]: prob want this to be an `error!` 😄

File: src/cluster/remote_handle.rs [hunk 3]
```diff
@@ -39,25 +59,31 @@ impl IndexHandle for RemoteIndex {
...
+                    e.into()
+                })
+        });
 
-        Box::new(fut)
+        Box::new(future::join_all(fut))
```
Comment (LucioFranco) [inline src/cluster/remote_handle.rs]: This actually has a concrete type so we don't have to box here but I have some ideas around this and its not a big deal right now but down the road we want to remove all these boxes.

File: src/cluster/rpc_server.rs [hunk 4]
```diff
@@ -84,73 +84,99 @@ impl RpcServer {
...
+    type PlaceReplicaFuture = future::FutureResult<Response<ResultReply>, Error>;
     type SearchIndexFuture = future::FutureResult<Response<SearchReply>, Error>;
 
-    fn place_index(&mut self, _request: Request<PlaceRequest>) -> Self::PlaceIndexFuture {
-        unimplemented!()
-    }
-
     fn list_indexes(&mut self, _: Request<ListRequest>) -> Self::ListIndexesFuture {
-        if let Ok(ref mut cat) = self.catalog.read() {
+        if let Ok(ref cat) = self.catalog.read() {
```
Comment (LucioFranco) [inline src/cluster/rpc_server.rs]: does this call block? if so we should probably wrap this call in a `blocking` https://docs.rs/tokio-threadpool/0.1.11/tokio_threadpool/fn.blocking.html

File: src/cluster/rpc_server.rs [hunk 5]
```diff
@@ -84,73 +84,99 @@ impl RpcServer {
...
-        if let Ok(ref mut cat) = self.catalog.read() {
-            let index = cat.get_index(&inner.index).unwrap();
-            let query: query::Request = serde_json::from_slice(&inner.query).unwrap();
-            match index.search_index(query) {
-                Ok(query_results) => {
-                    let query_bytes: Vec<u8> = serde_json::to_vec(&query_results).unwrap();
-                    let result = Some(RpcServer::create_result(0, "".into()));
-                    let resp = Response::new(RpcServer::create_search_reply(result, query_bytes));
-                    future::finished(resp)
+        if let Ok(ref cat) = self.catalog.read() {
```
Comment (LucioFranco) [inline src/cluster/rpc_server.rs]: same here

File: src/cluster/rpc_server.rs [hunk 6]
```diff
@@ -84,73 +84,99 @@ impl RpcServer {
...
+            Self::error_response(
+                Code::NotFound,
+                format!("Could not obtain lock on catalog for index: {}", inner.index),
+            )
+        }
+    }
+
+    fn place_index(&mut self, request: Request<PlaceRequest>) -> Self::PlaceIndexFuture {
+        let PlaceRequest { index, schema } = request.into_inner();
+        if let Ok(ref mut cat) = self.catalog.write() {
```
Comment (LucioFranco) [inline src/cluster/rpc_server.rs]: same here

File: src/handlers/search.rs (hunk 1 above)
Comment (LucioFranco) [inline src/handlers/search.rs]: You probably want to use `Runtime::block_on` one of the things I have been meaning to add to tokio is a `tokio_test::block_on` that should make this a bit easier.

File: src/index.rs [hunk 7]
```diff
@@ -86,39 +107,55 @@ impl IndexCatalog {
...
+        &mut self.local_handles
     }
 
     pub fn exists(&self, index: &str) -> bool {
-        self.get_collection().contains_key(&index.to_string())
+        self.get_collection().contains_key(index)
+    }
+
+    pub fn remote_exists(&self, index: &str) -> bool {
+        self.get_remote_collection().lock().unwrap().contains_key(index)
```
Comment (LucioFranco) [inline src/index.rs]: probably want to use `lock()?` like you do below

File: src/index.rs [hunk 8]
```diff
@@ -144,35 +181,51 @@ impl IndexCatalog {
...
 
-    pub fn search_index(&self, index: &str, search: Request) -> Result<SearchResults> {
-        match self.get_index(index) {
-            Ok(hand) => hand.search_index(search),
-            Err(e) => Err(e),
+    pub fn search_local_index(&self, index: &str, search: Request) -> impl Future<Item = Vec<SearchResults>, Error = Error> + Send {
+        dbg!(&search);
+        self.get_index(index)
+            .and_then(move |hand| hand.search_index(search).map(|r| vec![r]))
+            .into_future()
```
Comment (LucioFranco) [inline src/index.rs]: I believe you don't need this into future here

File: src/index.rs [hunk 9]
```diff
@@ -213,15 +267,25 @@ pub mod tests {
...
+                for idx in i.1 {
+                    println!("IDX={}", idx);
+                }
+                future::ok(())
+            })
+            .map_err(|_| ());
+        let s = service.select(reff);
 
-        tokio::run(s);
+        rt.block_on(s);
```
Comment (LucioFranco) [inline src/index.rs]: yeah this is what `tokio_test::block_on` will do pretty much. though I think without the `shutdown_on_idle` but `shutdown_now`.

File: src/cluster/remote_handle.rs (hunk 3 above)
Comment (hntd187) [reply src/cluster/remote_handle.rs]: so you'd want the return type to be a `JoinAll` ?

Comment (LucioFranco) [reply src/cluster/remote_handle.rs]: We can, that would remove the need for a `Box` call. though may be harder than that if we use the combinators, though I wouldnt worry about this now. I have some ideas how we can improve this in the future.

**Pull request merged**